llvm/lib/Target/RISCV/RISCVISelLowering.cpp

```diff
@@ -20130,6 +20105,135 @@ unsigned RISCVTargetLowering::getMinimumJumpTableEntries() const {
   return Subtarget.getMinimumJumpTableEntries();
 }
 
+void RVVArgDispatcher::constructHelper(Type *Ty) {
+  const DataLayout &DL = MF->getDataLayout();
+  const Function &F = MF->getFunction();
+  LLVMContext &Context = F.getContext();
+
+  StructType *STy = dyn_cast<StructType>(Ty);
+  if (STy && STy->containsHomogeneousScalableVectorTypes()) {
+    Type *ElemTy = STy->getTypeAtIndex(0U);
+    EVT VT = TLI->getValueType(DL, ElemTy);
+    MVT RegisterVT =
+        TLI->getRegisterTypeForCallingConv(Context, F.getCallingConv(), VT);
+
+    RVVArgInfos.push_back({STy->getNumElements(), RegisterVT, false});
+  } else {
+    SmallVector<EVT, 4> ValueVTs;
+    ComputeValueVTs(*TLI, DL, Ty, ValueVTs);
+
+    for (unsigned Value = 0, NumValues = ValueVTs.size(); Value != NumValues;
+         ++Value) {
+      EVT VT = ValueVTs[Value];
+      MVT RegisterVT =
+          TLI->getRegisterTypeForCallingConv(Context, F.getCallingConv(), VT);
+      unsigned NumRegs =
+          TLI->getNumRegistersForCallingConv(Context, F.getCallingConv(), VT);
+
+      // Skip non-RVV register type
+      if (!RegisterVT.isVector())
+        continue;
+
+      if (RegisterVT.isFixedLengthVector())
+        RegisterVT = TLI->getContainerForFixedLengthVector(RegisterVT);
+
+      RVVArgInfo Info{1, RegisterVT, false};
+
+      while (NumRegs--)
+        RVVArgInfos.push_back(Info);
+    }
+  }
+}
+
+void RVVArgDispatcher::construct(bool IsRet) {
+  const Function &F = MF->getFunction();
+
+  if (IsRet)
+    constructHelper(F.getReturnType());
+  else if (CLI)
+    for (const TargetLowering::ArgListEntry &Arg : CLI->getArgs())
+      constructHelper(Arg.Ty);
+  else if (GISelCLI)
+    for (const CallLowering::ArgInfo &Arg : GISelCLI->OrigArgs)
+      constructHelper(Arg.Ty);
+  else
+    for (const Argument &Arg : F.args())
+      constructHelper(Arg.getType());
+
+  for (auto &Info : RVVArgInfos)
+    if (Info.NF == 1 && Info.VT.getVectorElementType() == MVT::i1) {
+      Info.FirstVMask = true;
+      break;
+    }
+}
+
+void RVVArgDispatcher::allocatePhysReg(unsigned NF, unsigned LMul,
+                                       unsigned StartReg) {
+  assert((StartReg % LMul) == 0 &&
+         "Start register number should be multiple of lmul");
+  const MCPhysReg *VRArrays;
+  switch (LMul) {
+  default:
+    report_fatal_error("Invalid lmul");
+  case 1:
+    VRArrays = ArgVRs;
+    break;
+  case 2:
+    VRArrays = ArgVRM2s;
+    break;
+  case 4:
+    VRArrays = ArgVRM4s;
+    break;
+  case 8:
+    VRArrays = ArgVRM8s;
+    break;
+  }
+
+  for (unsigned i = 0; i < NF; ++i)
+    if (StartReg)
+      AllocatedPhysRegs.push_back(VRArrays[(StartReg - 8) / LMul + i]);
+    else
+      AllocatedPhysRegs.push_back(0);
+}
+
+// This function determines if each RVV argument is passed by register.
```

Comment:
It looks like this function has the side effect of allocating registers for arguments that are determined to be passed by register. Should we include that in the docstring?